web-api/src/extractor.rs

```diff
@@ -0,0 +1,239 @@
+    collections::HashSet,
+    hash::{Hash, Hasher},
+};
+
+use derive_more::Display;
+use mime::{Mime, Name};
+use mime_serde_shim::Wrapper as SerDeMime;
+use reqwest::Client;
+use serde::{Deserialize, Serialize};
+use tracing::info;
+use url::Url;
+
+use crate::{
+    error::common::{FileUploadNotEnabled, InvalidBinary},
+    ingestion::preprocessor::PreprocessError,
+    models::DocumentSnippet,
+};
+
+#[derive(Debug, Deserialize, Serialize, Clone)]
+pub struct Config {
+    /// File text extraction is available.
+    #[serde(default)]
+    pub(crate) enabled: bool,
+    #[serde(flatten)]
+    config: ExtractorConfig,
+}
+
+#[derive(Debug, Deserialize, Serialize, Clone)]
+#[serde(tag = "extractor")]
+pub enum ExtractorConfig {
+    #[serde(rename = "tika")]
+    Tika {
+        /// Tika server to contact
+        #[serde(default = "default_text_extractor_url")]
+        url: String,
+        /// Allowed content-type. If empty everything is allowed.
+        #[serde(default)]
+        allowed_content_type: Vec<SerDeMime>,
+    },
+}
+
+fn default_text_extractor_url() -> String {
+    "http://localhost:9998".into()
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            enabled: false,
+            config: ExtractorConfig::Tika {
+                url: default_text_extractor_url(),
+                allowed_content_type: Vec::new(),
+            },
+        }
+    }
+}
+
+#[derive(Debug, Deserialize)]
+struct TikaResponse {
+    #[serde(rename = "Content-Type")]
+    content_type: String,
+
+    #[serde(rename = "X-TIKA:content")]
+    content: Option<String>,
+}
+
+// We want to compare only the type and subtype
+#[derive(Debug, Display)]
+struct CmpMime(Mime);
+
+impl CmpMime {
+    fn project(&self) -> (Name<'_>, Name<'_>) {
+        (self.0.type_(), self.0.subtype())
+    }
+}
+
+impl PartialEq for CmpMime {
+    fn eq(&self, other: &Self) -> bool {
+        self.cmp(other).is_eq()
+    }
+}
+impl Eq for CmpMime {}
+
+impl PartialOrd for CmpMime {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for CmpMime {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.project().cmp(&other.project())
+    }
+}
+
+impl Hash for CmpMime {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.project().hash(state);
+    }
+}
+
+/// Extract text and other information from different format
+pub(crate) struct TextExtractor {
+    inner: ExtractorInner,
+}
+
+impl TextExtractor {
+    pub(crate) fn new(config: &Config) -> Result<Self, anyhow::Error> {
+        if !config.enabled {
+            return Ok(Self {
+                inner: ExtractorInner::Disabled,
+            });
+        }
+
+        let inner = match &config.config {
+            ExtractorConfig::Tika {
+                url,
+                allowed_content_type,
+            } => {
+                let url = Url::parse(url)?;
+                if url.cannot_be_a_base() {
```

Comment:
In `web-api-shared` there is `SegmentableUrl`  which is a `Url` newtype you could use here, through weather you want to use this is fully your choice I won't request a change for it.